From 852916982b5f201653921cd5efd65f2e1af37754 Mon Sep 17 00:00:00 2001
From: Kazunori Kobayashi <kkobayas@igel.co.jp>
Date: Wed, 12 Jun 2013 18:00:37 +0900
Subject: [PATCH 602/611] ARM: shmobile: r8a7778: add a register region for an
 ICB errata workaround to UIO

This patch allows to access a register to control ICB cache for
an ICB errata workaround from userland. With the register, the ICB
cache can be set to being enabled or disabled, and flushed as required.

Signed-off-by: Yusuke Goda <yusuke.goda.sx@renesas.com>
---
 arch/arm/mach-shmobile/board-bockw.c          |  8 ++++++++
 arch/arm/mach-shmobile/include/mach/r8a7778.h |  1 +
 arch/arm/mach-shmobile/setup-r8a7778.c        | 20 ++++++++++++++++++++
 3 files changed, 29 insertions(+)

diff --git a/arch/arm/mach-shmobile/board-bockw.c b/arch/arm/mach-shmobile/board-bockw.c
index f3847ca..56fd51e 100644
--- a/arch/arm/mach-shmobile/board-bockw.c
+++ b/arch/arm/mach-shmobile/board-bockw.c
@@ -195,6 +195,13 @@ static struct uio_info meram_plat = {
 	.irq = -1,
 };
 
+/* ICB Cache Control */
+static struct uio_info icb_plat = {
+	.name = "ICB_CACHE_CTRL",
+	.version = "0",
+	.irq = -1,
+};
+
 /* SSI/AK4554 */
 static struct rsnd_ssi_platform_info rsnd_ssi[] = {
 	RSND_SSI_UNUSED, /* SSI 0 */
@@ -495,6 +502,7 @@ static void __init bockw_init(void)
 	r8a7778_add_meram_device(&meram_plat);
 	r8a7778_add_vio_device();
 	r8a7778_add_vpu_device();
+	r8a7778_add_icb_device(&icb_plat);
 	platform_device_register_data(&platform_bus, "soc-camera-pdrv", 0,
 				      &iclink0_ml86v7667,
 				      sizeof(iclink0_ml86v7667));
diff --git a/arch/arm/mach-shmobile/include/mach/r8a7778.h b/arch/arm/mach-shmobile/include/mach/r8a7778.h
index 1b61885..7a1a9c6 100644
--- a/arch/arm/mach-shmobile/include/mach/r8a7778.h
+++ b/arch/arm/mach-shmobile/include/mach/r8a7778.h
@@ -46,6 +46,7 @@ extern void r8a7778_add_vpc_device(struct uio_info *info);
 extern void r8a7778_add_meram_device(struct uio_info *info);
 extern void r8a7778_add_vio_device(void);
 extern void r8a7778_add_vpu_device(void);
+extern void r8a7778_add_icb_device(struct uio_info *info);
 
 
 extern void r8a7778_init_late(void);
diff --git a/arch/arm/mach-shmobile/setup-r8a7778.c b/arch/arm/mach-shmobile/setup-r8a7778.c
index d2f528f..7ff4c54 100644
--- a/arch/arm/mach-shmobile/setup-r8a7778.c
+++ b/arch/arm/mach-shmobile/setup-r8a7778.c
@@ -562,6 +562,26 @@ void __init r8a7778_add_vpu_device(void)
 	platform_device_register(&vpu_device);
 }
 
+/* ICB Cache Control */
+static struct resource icb_resources[] __initdata = {
+	DEFINE_RES_MEM(0xfd050000, 0x98),
+};
+
+void __init r8a7778_add_icb_device(struct uio_info *info)
+{
+	struct platform_device_info pdevinfo = {
+		.parent		= &platform_bus,
+		.name		= "uio_pdrv_genirq",
+		.id		= 3,
+		.res		= icb_resources,
+		.num_res	= ARRAY_SIZE(icb_resources),
+		.data		= info,
+		.size_data	= sizeof(*info),
+	};
+
+	platform_device_register_full(&pdevinfo);
+}
+
 void __init r8a7778_add_standard_devices(void)
 {
 	int i;
-- 
1.8.2

